### ISSUE TITLE
[question] Logging in brigade w/ structured return value

### ISSUE THREAD

**Issue by FreakTheMighty** [brigadecore/brigade#1055]
What sort of approach is suggested if you'd like your jobs to return a structured JSON response, but would still like to benefit from logging outputs for inspection in kashti?

At the moment, the option I can think of is to write the results to a file and reserve stdout for logging. This seems workable, but I don't love it. Getting the results directly from brigade by awaiting a job feels far more clean and would likely be less error prone than designing my own results store.

Is there anyway to use a different transport protocol, such as HTTP to avoid these sorts of collisions? I could send results back in an HTTP request, and continue to use stdout for logs.

**Comment by krancour** [comment]
> the option I can think of is to write the results to a file and reserve stdout for logging. This seems workable, but I don't love it.

Could you explain why you're averse to that? I would, in fact, think it _unusual_ for results to be returned via logs. To demonstrate, consider a common use case-- using a job to run tests or execute a build. In the case of tests, the coverage report may be the artifact that is produced. In the case of a build, the artifact produced by the job might be a binary, a JAR, or a Docker image. In either case, these artifacts don't usually live in the logs, right? In the case of coverage reports, they're be uploaded to someplace like codecov.io or in the case of a binary, published to object storage, perhaps. In the case of a Docker image, it might be pushed to DockerHub or some other Docker registry.

I'm not sure what it is that your job does, but burying the output of the job in the logs does not strike me as advisable.

That said... if you really wanted to do this, you could consider using fluentd (or fluent-bit) to aggregate logs from the job(s) in question.

> Is there anyway to use a different transport protocol, such as HTTP to avoid these sorts of collisions?

I'm not sure how you perceive this working. Taking a guess here-- you want jobs to listen for HTTP requests so they can report their status? It would work, with several caveats, with the first being that this would only work for as long as the job was running. Jobs are, by definition, not long-running affairs. They're meant to run to completion. If you stop a job from exiting, the build that job is part of will, itself, never complete. The second caveat is that you'd have to go behind brigade's back so-to-speak to define a k8s service that selects pods from the job(s) in question in order to get at this info over HTTP. That's _doable_, but you quickly find yourself in a sticky situation and scratching your head at what's going to happen if two builds are running concurrently, running two separate copies of the same job-- the selector will select both and you won't know which one the results you get back belong to.

Bottom line... I would strongly encourage you to treat job results as an artifact and conclude your job by pushing those results somewhere.

**Comment by FreakTheMighty** [comment]
@krancour thank you for that thorough explanation. Here's more context about where I'm coming from.

First, I was inspired to return structured data by the brigade.sh docs [here](https://docs.brigade.sh/topics/scripting/#jobs-and-return-values). I'm not sure if this is _unsual_, but seems reasonable to me. The docs note:

>With carefully constructed containers, you can get sophisticated and send structured data like JSON from one job to another. But remember that what is captured is the standard output (STDOUT) of the job, which is often where log data will also be sent. Sometimes it makes more sense to write structured files to the shared storage instead.

This is exactly the use I have, an upstream job's results should be passed to downstream jobs. If I understand correctly, bridage.sh would make STDOUT available as the "results" using the API like so:
```javascript
one.run().then( result => {
  two.tasks = ["echo hello " + result.toString()]
   two.run().then( result2 => {
    console.log(result2.toString())
  })
})
```
I would also think that STDOUT would be present in the logs. It seems to me that these two uses of STDOUT, logging and returning structured results, won't work well together. The issues I see:
 * You would want logging to happen through out a job, not just at the end
 * You'd likely want to log more than you'd typically want in a return payload

I don't have a *strong* aversion to writing to a "results" file  . . . but it feels like it is something brigade's API at least partially attempts to provide for. Being able to do `result = await jobProm` is cleaner than:
```javascript
const tmpResults1 = createResultsPath();
job.tasks = [`cmd --out-json ${tmpResults1 }`]
await job.run();
cosnst results = readResults(tmpResults1);
```
What do you think the use cases for having a job return results?

----

What I was thinking in terms of using HTTP is inspired by how knative and openfaas works. Rather than having jobs accept input and write output via STDIN/OUT, it is done via HTTP. This means that you can send and retrieve structured data, while logging however you'd like.

I believe openfaas started by only supporting STDIN/OUT, but have [built a next generation](https://github.com/openfaas-incubator/of-watchdog) that can communicate with their "functions" via http. \

**Comment by krancour** [comment]
The docs demonstrate the range of what _could_ be done. Something I've seen done frequently is capture the logs from one job and feed them into a second job whose entire role is to push them somewhere. So yes... there's a mechanism provided for doing this sort of thing, but I think discretion is advised. Indeed, I think this last bit of the statement is quite relevant here:

> Sometimes it makes more sense to write structured files to the shared storage instead.

As for this:

```
const tmpResults1 = createResultsPath();
job.tasks = [`cmd --out-json ${tmpResults1 }`]
await job.run();
cosnst results = readResults(tmpResults1);
```
Presumably you feed the results into the next job? If so, this appears more complicated that it needs to be.  Brigade offers optional per-build shared storage that can be mounted into your jobs. You can have the first job write results to any path in the shared storage and have the second job read them from the same location.

**Comment by FreakTheMighty** [comment]
@krancour got it. Sounds like I'm trying to use brigade in a way that it is not intended to be used. Thanks for taking the time to answer my questions.

---
You're right about being able to feed a path directly into the next job. There are times where I would want the results in the brigade.js to help determine how exactly down stream jobs are orchestrated. Still, the answer is, "use files on disk."

**Comment by krancour** [comment]
> There are times where I would want the results in the brigade.js to help determine how exactly down stream jobs are orchestrated.

That is an interesting thought. I'm going to contemplate that because... currently the build-level shared storage is provisioned by the worker itself toward the beginning of its execution. This means its not mounted _into_ the worker.

I am wondering if in 2.0, there could be a benefit to shared storage being provisioned by the controller instead, then mounted into the worker, creating conditions where the worker could directly access artifacts created by jobs such that, in your case, the worker could intelligently make orchestration decisions based on structured results from the jobs without relying on the jobs' logs.

**Comment by FreakTheMighty** [comment]
Alright, then in an effort to beat a dead 🏇 ... would you consider supporting an HTTP transport similar to knative or openfaas? Not sure you need the orchestrator to have access to shared storage if you could more easily return structured data.

Just imagine how tidy an brigade.js would be if you could simply `const results = await job.run()`, and get a JSON body.

**Comment by krancour** [comment]
@FreakTheMighty can you please elaborate on how you imagine that working? I'm not sure that my understanding of the suggestion is necessarily correct.

**Comment by FreakTheMighty** [comment]
Perhaps brigade could support an HTTP flow like this. We specify an image, just like we do now, but now brigade passes a `PORT` env variable that our image must listen on for an HTTP request. Now, instead of passing tasks as a list of strings, they could be JSON serializable objects. These would then get POSTed to our image and the response returned to the brigade.js.

Again, this is heavily inspired by my experience with Knative. Check out these example https://cloud.google.com/run/docs/quickstarts/build-and-deploy

The benefits from my perspective:

* Writing arg parsers for tasks that will only be used by a machine doesn't feel as clean as defining a JSON structure
* As mentioned earlier, using HTTP for the communication protocol between brigade and tasks frees STDIO/OUT up to just perform logging. Using STDOUT for returning structured data is error prone, any errant print statement could break the output.
* Using intermediate files would work fine if they were accessible in the brigade.js script, however I feel getting structured return data would be a powerful feature. Why make each dev/script re-implement such a pattern?

**Comment by krancour** [comment]
@FreakTheMighty I'm quite familiar with knative and I think there is a fundamental difference between how they model functions and how brigade models jobs. A knative function may receive input and output over HTTP, but it is also "long running" in the sense that a function can serve multiple requests in the time it is up, prior to eventually being scaled to zero as a result of inactivity. By contrast, a brigade worker executes _one_ discrete pipeline, which may fan out to many discrete jobs. They are "short-lived," running to completion by processing _one and only one_ workload, and then exiting. (It would be more appropriate to compare brigade jobs to kubernetes Jobs than to knative functions.)

If you'd like us to change that, you're asking for a _radical_ change to brigade's architecture. That aside, this would also significantly raise the bar for containers used to implement a job. They'd have to listen for HTTP requests? How much sense does it make for the ubiquitous case of executing some tests or compiling some code-- jobs that could otherwise be implemented using _existing_ containers for the language of your choice, but couldn't be if we add in this requirement?

I _do_ still recognize that we could make machine-pasrseable results more readily available to the worker by mounting shared build storage to the worker. I will make sure that makes its way into the 2.0 proposal.
